bundles/io/org.eclipse.smarthome.io.voice/src/main/java/org/eclipse/smarthome/io/voice/text/AbstractRuleBasedInterpreter.java

```diff
@@ -106,23 +107,89 @@ public String interpret(Locale locale, String text) throws InterpretationExcepti
     }
 
     private void invalidate() {
-        identifierTokens = null;
+        allItemTokens = null;
+        itemTokens = null;
         languageRules = null;
     }
 
     /**
      * All the tokens (name parts) of the names of all the items in the {@link ItemRegistry}.
      *
+     * @param locale The locale that is to be used for preparing the tokens.
      * @return the identifier tokens
      */
-    HashSet<String> getIdentifierTokens() {
-        if (identifierTokens == null) {
-            identifierTokens = new HashSet<String>();
+    HashSet<String> getAllItemTokens(Locale locale) {
+        if (allItemTokens == null) {
+            allItemTokens = new HashMap<Locale, HashSet<String>>();
+        }
+        HashSet<String> localeTokens = allItemTokens.get(locale);
+        if (localeTokens == null) {
+            allItemTokens.put(locale, localeTokens = new HashSet<String>());
             for (Item item : itemRegistry.getAll()) {
-                identifierTokens.addAll(splitName(item.getLabel(), true));
+                localeTokens.addAll(tokenize(locale, item.getLabel()));
+            }
+        }
+        return localeTokens;
+    }
+
+    /**
+     * Retrieves the list of identifier token sets per item currently contained in the {@link ItemRegistry}.
+     *
+     * @param locale The locale that is to be used for preparing the tokens.
+     * @return the list of identifier token sets per item
+     */
+    HashMap<Item, ArrayList<HashSet<String>>> getItemTokens(Locale locale) {
+        if (itemTokens == null) {
+            itemTokens = new HashMap<Locale, HashMap<Item, ArrayList<HashSet<String>>>>();
+        }
+        HashMap<Item, ArrayList<HashSet<String>>> localeTokens = itemTokens.get(locale);
+        if (localeTokens == null) {
+            itemTokens.put(locale, localeTokens = new HashMap<Item, ArrayList<HashSet<String>>>());
+            for (Item item : itemRegistry.getItems()) {
+                if (item.getGroupNames().isEmpty()) {
+                    addItem(locale, localeTokens, new HashSet<String>(), item);
+                }
+            }
+        }
+        return localeTokens;
+    }
+
+    private void addItem(Locale locale, HashMap<Item, ArrayList<HashSet<String>>> target, HashSet<String> tokens,
+            Item item) {
+        HashSet<String> nt = new HashSet<String>(tokens);
+        nt.addAll(tokenize(locale, item.getLabel()));
+        ArrayList<HashSet<String>> list = target.get(item);
+        if (list == null) {
+            target.put(item, list = new ArrayList<HashSet<String>>());
+        }
+        list.add(nt);
+        if (item instanceof GroupItem) {
+            for (Item member : ((GroupItem) item).getMembers()) {
+                addItem(locale, target, nt, member);
+            }
+        }
```

Comment:
Items can be members of many groups. This is the reason why the list is required. It stores every possible "path" to the item. Yes, it is correct, that the token sequence represents the whole path to the item (containing also all the parents and their tokens). I'll add some documentation.